[SEC-11068] Add Azure and GCP to PA Documentation (#20591)
DataDog/documentation: content/en/security/cloud_security_management/guide/public-accessibility-logic.md

For more information on network reachability, see the [AWS documentation][34] and the [AWS Network Reachability Analyser][35]. Currently, the `Is Publicly Accessible` facet is only available for AWS resources.

> This page details on a resource-by-resource basis the logic behind public accessibility determination for a resource.

Reviewer: ```suggestion
```

Reviewer: I would remove this line as it's a duplicate of the second sentence in the initial paragraph.

> **Note:** Not all resources are able to be classified as publicly accessible or not publicly accessible through only their resource configuration. Such resources are not included here.

Reviewer 2: not clear what this means

## Resource dependency graph

> The following diagrams show how related resources are used to determine whether other resources are publicly accessible. For example, a CloudTrail Trail stored in a public S3 bucket is itself publicly accessible. If a resource is publicly accessible because of another resource, the relationship is shown in the Cloud Security Management Misconfigurations Resource relationships graph.

Reviewer 2: ```suggestion
The following diagrams show how related resources are used to determine whether other resources are publicly accessible. For example, an AWS CloudTrail Trail stored in a public Amazon S3 bucket is itself publicly accessible. If a resource is publicly accessible because of another resource, the relationship is shown in the Cloud Security Management Misconfigurations resource relationships graph.
```

Use this diagram as a reference for how resources are correlated to determine public accessibility. 

**Note**: Not all resources with the Publicly Accessible attribute are shown in these diagrams.

### AWS

> {{< img src="security/cloud_security_management/guide/public_accessibility_relationships_aws.png" alt="A graph diagram showing the relationships between resources that are used to determine public accessibility for AWS" width="100%">}}

Reviewer 2: doesn't seem to be part of this PR?

Reviewer: Noting that I see the same.

### Azure

> {{< img src="security/cloud_security_management/guide/public_accessibility_relationships_azure.png" alt="A graph diagram showing the relationships between resources that are used to determine public accessibility for Azure" width="70%">}}

Reviewer 2: doesn't seem to be part of this PR?

Reviewer: Noting that I see the same.

> ### GCP

Reviewer 2: We're supposed to call it "Google Cloud" now, but not sure what's our stance on the docs side, I'll defer to someone from the documentation team

Reviewer: Agreed. We should change this to "Google Cloud".

> ### GCP

Reviewer: ```suggestion
### Google Cloud
```

> {{< img src="security/cloud_security_management/guide/public_accessibility_relationships_gcp.png" alt="A graph diagram showing the relationships between resources that are used to determine public accessibility for GCP" width="50%">}}

Reviewer 2: doesn't seem to be part of this PR?

Reviewer: Noting that I see the same.

## Resource dependency graph

The following diagrams show how related resources are used to determine whether other resources are publicly accessible. For example, an AWS CloudTrail Trail stored in a public Amazon S3 bucket is itself publicly accessible. If a resource is publicly accessible because of another resource, the relationship is shown in the Cloud Security Management Misconfigurations resource relationships graph.

> Use this diagram as a reference for how resources are correlated to determine public accessibility. 

Reviewer: ```suggestion
```

## Google Cloud Public accessibility logic by resource

### Google Cloud Compute Firewall

A Compute Firewall (`gcp_compute_firewall`) grants public access if:

| Criteria | Explanation |
|----------|-------------|
|The firewall has one or more rules whose protocol is TCP or all and which have `0.0.0.0/0` or `::/0` in their `source_ranges`. | These CIDR prefixes allow access from the Internet, and are the protocol values that are relevant for determining public access. |
|The firewall's direction is `ingress`. | This means that the firewall is relevant for inbound access from the Internet. |

For more information about using Compute Firewalls, [Choose to allow or disallow blob public access on Azure Storage accounts][47].

### Google Cloud Compute Instance

A Compute Instance (`gcp_compute_instance`) is considered publicly accessible if:

| Criteria | Explanation |
|----------|-------------|
|The compute instance has a public IP address, meaning at least one of its network interfaces has a public IP address defined in its access configurations, | To learn more about adding an external IP to a compute instance, see [Reserve a static external IP address][48]. |

> |The compute instance has associated firewall rules that combine to open some range of ports to the internet. The firewall rules can be associated with the instance by:<br><p><ul><li>Having no `target_tags` or `target_service_accounts`, meaning the rule applies to the whole network.</li><li>Having `target_service_accounts` associated with one of the compute instance’s `service_accounts`.</li><li>Having some `target_tags` that match the compute instance’s network tags.</li></ul></p>The rules should grant public access (see [Google Cloud Compute Firewall](#google-cloud-compute-firewall)). | To learn how compute firewall rules are used to restrict port ranges for a compute instance, see [Firewall rule components][49]. |

Reviewer: ```suggestion
|The compute instance has associated firewall rules that combine to open some range of ports to the internet. The firewall rules can be associated with the instance by:<br><p><ul><li>Having no `target_tags` or `target_service_accounts`, meaning the rule applies to the whole network.</li><li>Having `target_service_accounts` associated with one of the compute instance's `service_accounts`.</li><li>Having some `target_tags` that match the compute instance's network tags.</li></ul></p>The rules should grant public access (see [Google Cloud Compute Firewall](#google-cloud-compute-firewall)). | To learn how compute firewall rules are used to restrict port ranges for a compute instance, see [Firewall rule components][49]. |
```

See [Amazon SQS security best practices][33] for more information about public SQS queues.

> ## Azure Public accessibility logic by resource

Reviewer: ```suggestion
## Azure public accessibility logic by resource
```

### Azure Network Security Group (NSG)

An Azure NSG (`azure_security_group`) grants public access if:

| Criteria | Explanation |
|----------|-------------|

> |The security group has rules with protocol `tcp`, `udp` or `*` | These are the protocol values that are relevant for determining public access for the Azure resources that follow |

Reviewer: ```suggestion
|The security group has rules with protocol `tcp`, `udp` or `*`. | These are the protocol values that are relevant for determining public access for Azure resources. |
```

> |The security group has `inbound` rules with access set to `Allow` | Having these values indicates that the rule is allowing inbound traffic |

Reviewer: ```suggestion
|The security group has `inbound` rules with access set to `Allow`. | These values indicates that the rule is allowing inbound traffic. |
```

> |The security group has rules with source_address_prefix equal to `*`, `0.0.0.0`, `/0`, `::/0`, `internet`, or `any` | These CIDR prefixes allow access to the internet |

Reviewer: ```suggestion
|The security group has rules with source_address_prefix equal to `*`, `0.0.0.0`, `/0`, `::/0`, `internet`, or `any`. | These CIDR prefixes allow access to the internet. |
```

> |The rules which match the above properties combine with any other `Deny` rules of higher priority to open at least one port to the internet | See [here][39] for how Azure combines security group rules to calculate access |

Reviewer: ```suggestion
|The rules which match the above properties combine with any other `Deny` rules of higher priority to open at least one port to the Internet. | See [Security rules][39] to learn how Azure combines security group rules to calculate access. |
```

> For more details on how Azure NSGs allow and deny internet access for a resource, see [here][40].

Reviewer: ```suggestion
For details on how Azure NSGs allow and deny Internet access for a resource, see [Network Security Groups][40].
```

### Azure Virtual Machine Instance

A Virtual Machine Instance (`azure_virtual_machine_instance`) is considered publicly accessible if:

* _Network Security Group allowing public access:_

| Criteria | Explanation |
|----------|-------------|

> |The virtual machine instance has a public IP address attached to its network interface | A public IP is required for internet access to a virtual machine instance |

Reviewer 2: ```suggestion
|The virtual machine instance has a public IP address attached to one of its network interfaces | A public IP is required for internet access to a virtual machine instance |
```

> |The virtual machine instance has a network security group granting public access attached to its network interface | Refer to the Azure Network Security Group (NSG) section for how a network can grant public access |

Reviewer 2: ```suggestion
|The virtual machine instance has a network security group granting public access attached to one of its network interfaces | Refer to the Azure Network Security Group (NSG) section for how a network can grant public access |
```

***OR***

* _Public IP with SKU "Basic":_

| Criteria | Explanation |
|----------|-------------|

> |The virtual machine instance has a public IP address with SKU Basic attached to its network interface | A public IP address with SKU basic is open by default (see [here][41]) |

Reviewer: ```suggestion
|The virtual machine instance has a public IP address with SKU Basic attached to its network interface. | A public IP address with SKU basic is open by default (see [Public IP addresses][41]). |
```

> |The virtual machine instance has no attached network security groups | If no network security groups are attached, then there are no rules blocking access through the open public IP address |

Reviewer: ```suggestion
|The virtual machine instance has no attached network security groups. | If no network security groups are attached, then there are no rules blocking access through the open public IP address. |
```

> Learn more about Azure Virtual Machine Instances and public access [here][42].

Reviewer: ```suggestion
To learn more about Azure Virtual Machine Instances and public access, see [Associate a public IP address to a virtual machine][42].
```

### Azure Managed Disk

> A Managed Disk (`azure_managed_disk`) is considered publicly accessible if:

Reviewer 2: let's chat about this one (it's about business logic not docs) but I don't think it's fair/correct to consider such a disk as "publicly accessible"

| Criteria | Explanation |
|----------|-------------|

> |The managed disk’s `network_access_policy` is set to `AllowAll` | This setting means that “The disk can be exported or uploaded to from any network” (see [here][43]) |

Reviewer: ```suggestion
|The managed disk’s `network_access_policy` is set to `AllowAll`. | This setting means that the disk can be exported or uploaded to from any network (see [Disks - Create Or Update][43]). |
```

> |The managed disk’s virtual machine instance is publicly accessible | Refer to the Azure Virtual Machine Instance section above for how a virtual machine is considered publicly accessible |

Reviewer: ```suggestion
|The managed disk's virtual machine instance is publicly accessible. | To learn how a virtual machine is considered publicly accessible, see [Azure Virtual Machine Instance](#azure-virtual-machine-instance).  |
```

> For more information about restricting import/export access for Azure Managed Disks, refer [here][44].

Reviewer: ```suggestion
For more information about restricting import/export access for Azure Managed Disks, see [Network Security Groups][44].
```

### Azure Storage Account

> A Storage Account (`azure_storage_account`) grants public access if:

Reviewer 2: for this one, IMHO this is an "internal attribute" that's used for public access determination of buckets (just like `aws_s3_account_public_access_block`) but shouldn't be documented, because it doesn't make a storage account public (it just doesn't _proactively block public access if a bucket in the storage account is set as public_)

| Criteria | Explanation |
|----------|-------------|

> |The storage account’s `allow_blob_public_access` attribute is set to `true`, or is nonexistent | This means that the account allows public internet access to Azure Blob Storage |

Reviewer: ```suggestion
|The storage account's `allow_blob_public_access` attribute is set to `true`, or is nonexistent. | This means that the account allows public Internet access to Azure Blob Storage. |
```

> See more about configuring anonymous read access with Azure Storage Accounts [here][45].

Reviewer: ```suggestion
To learn more about configuring anonymous read access with Azure Storage Accounts, see [Configure anonymous read access for containers and blobs][45].
```

### Azure Storage Blob Container

A Storage Blob Container (`azure_storage_blob_container`) is considered publicly accessible if:

| Criteria | Explanation |
|----------|-------------|

> |The storage blob container’s `public_access` attribute is not set to `None` | This means that the account allows public internet access to Azure Blob Storage |

Reviewer 2: double negation makes it hard to read, let's use one of the enum values from https://learn.microsoft.com/en-us/rest/api/storageservices/list-containers2?tabs=microsoft-entra-id#response-body which can be `blob` or `container`

> |The storage blob container’s `public_access` attribute is not set to `None` | This means that the account allows public internet access to Azure Blob Storage |

Reviewer: ```suggestion
|The storage blob container's `public_access` attribute is not set to `None`. | This means that the account allows public Internet access to Azure Blob Storage. |
```

|The storage blob container is part of a storage account that grants public access | Refer to the Azure Storage Account section above for how a storage account can grant public access |

> Learn more about disallowing blob public access on Azure Storage accounts [here][46].

Reviewer: ```suggestion
To learn more about disallowing blob public access on Azure Storage accounts, see [Choose to allow or disallow blob public access on Azure Storage accounts][46].
```

## GCP Public accessibility logic by resource

### GCP Compute Firewall

A Compute Firewall (`gcp_compute_firewall`) grants public access if:

| Criteria | Explanation |
|----------|-------------|

> |The firewall has one or more rules whose protocol is tcp or all and which have `0.0.0.0/0` or `::/0` in their `source_ranges` | These CIDR prefixes allow access from the internet, and these are the protocol values that are relevant for determining public access |

Reviewer: ```suggestion
|The firewall has one or more rules whose protocol is TCP or all and which have `0.0.0.0/0` or `::/0` in their `source_ranges`. | These CIDR prefixes allow access from the Internet, and are the protocol values that are relevant for determining public access. |
```

> |The firewall’s direction is `ingress` | This means that the firewall is relevant for inbound access from the internet |

Reviewer: ```suggestion
|The firewall's direction is `ingress`. | This means that the firewall is relevant for inbound access from the Internet. |
```

> Discover more information about using Compute Firewalls [here][47].

Reviewer: ```suggestion
For more information about using Compute Firewalls, [Choose to allow or disallow blob public access on Azure Storage accounts][47].
```

### GCP Compute Instance

A Compute Instance (`gcp_compute_instance`) is considered publicly accessible if:

| Criteria | Explanation |
|----------|-------------|

> |The compute instance has a public IP address, meaning at least one of its network interfaces has a public IP address defined in its access configurations | Learn more about adding an external IP to a compute instance [here][48] |

Reviewer: ```suggestion
|The compute instance has a public IP address, meaning at least one of its network interfaces has a public IP address defined in its access configurations, | To learn more about adding an external IP to a compute instance, see [Reserve a static external IP address][48]. |
```

> |The compute instance has associated firewall rules that combine to open some range of ports to the internet. The firewall rules can be associated with the instance by:<br><p><ul><li>Having no `target_tags` or `target_service_accounts`, meaning the rule applies to the whole network</li><li>Having `target_service_accounts` associated with one of the compute instance’s `service_accounts`</li><li>Having some `target_tags` that match the compute instance’s network tags.</li></ul></p>The rules should grant public access (see “GCP Compute Firewall” above) | See [here][49] for how compute firewall rules are used to restrict port ranges for a compute instance. |

Reviewer: ```suggestion
|The compute instance has associated firewall rules that combine to open some range of ports to the internet. The firewall rules can be associated with the instance by:<br><p><ul><li>Having no `target_tags` or `target_service_accounts`, meaning the rule applies to the whole network.</li><li>Having `target_service_accounts` associated with one of the compute instance’s `service_accounts`.</li><li>Having some `target_tags` that match the compute instance’s network tags.</li></ul></p>The rules should grant public access (see [GCP Compute Firewall](#gcp-compute-firewall)). | To learn how compute firewall rules are used to restrict port ranges for a compute instance, see [Firewall rule components][49]. |
```

Learn more about how compute firewall rules are used to restrict port ranges for a compute instance [here][50].

### GCP BigQuery Dataset

A BigQuery Dataset (`gcp_bigquery_dataset`) is considered publicly accessible if:

| Criteria | Explanation |
|----------|-------------|

> |The dataset has an IAM policy attached that has a `member` value of either `AllUsers` or `AllAuthenticatedUsers` | These members allow anyone on the internet to access the database. See more [here][51] |

Reviewer: ```suggestion
|The dataset has an IAM policy attached that has a `member` value of either `AllUsers` or `AllAuthenticatedUsers`. | These members allow anyone on the internet to access the database. See [IAM overview][51] for more information. |
```

> |The dataset has an IAM policy attached that binds it to one of the following roles: `roles/viewer`, `roles/owner`, `roles/editor`, `roles/bigquery.admin`, `roles/bigquery.metadataviewer`, `roles/bigquery.dataowner`, `roles/bigquery.dataeditor`, `roles/bigquery.dataviewer`, or `roles/bigquery.user`| These roles allow the person who accesses the resource to perform dangerous operations on the database. See the [role reference][52] for more information |

Reviewer: ```suggestion
|The dataset has an IAM policy attached that binds it to one of the following roles: `roles/viewer`, `roles/owner`, `roles/editor`, `roles/bigquery.admin`, `roles/bigquery.metadataviewer`, `roles/bigquery.dataowner`, `roles/bigquery.dataeditor`, `roles/bigquery.dataviewer`, or `roles/bigquery.user`. | These roles allow the person who accesses the resource to perform dangerous operations on the database. See the [role reference][52] for more information. |
```

> Explore more details about BigQuery Datasets [here][53].

Reviewer: ```suggestion
Learn more about [BigQuery Datasets][53].
```

### GCP Storage Bucket

A Storage Bucket (`gcp_storage_bucket`) is considered publicly accessible if:

| Criteria | Explanation |
|----------|-------------|
|The bucket has an IAM policy attached that has a `member` value of either `AllUsers` or `AllAuthenticatedUsers` | These members allow anyone on the internet to access the database. See more [here][54] |

> |The bucket has `public_access_prevention` set to `inherited` in its `iam_configuration` | This setting would block public access if it was set to `enforced`. See [here][55] for more information about the public access prevention setting.|

Reviewer: ```suggestion
|The bucket has `public_access_prevention` set to `inherited` in its `iam_configuration`. | This setting block public access if set to `enforced`. For more information about the public access prevention setting, see [Public access prevention][55]. |
```

[36]: https://docs.aws.amazon.com/vpc/latest/userguide/vpc-example-private-subnets-nat.html
[37]: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-listeners.html
[38]: https://docs.aws.amazon.com/elasticloadbalancing/latest/application/load-balancer-target-groups.html
[39]: https://learn.microsoft.com/en-us/azure/virtual-network/network-security-groups-overview#security-rules
[40]: https://azure.microsoft.com/fr-fr/blog/network-security-groups/
[41]: https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/public-ip-addresses
[42]: https://learn.microsoft.com/en-us/azure/virtual-network/ip-services/associate-public-ip-address-vm?tabs=azure-portal
[43]: https://learn.microsoft.com/en-us/rest/api/compute/disks/create-or-update?view=rest-compute-2023-04-02&tabs=HTTP
[44]: https://learn.microsoft.com/en-us/azure/virtual-machines/disks-enable-private-links-for-import-export-portal
[45]: https://learn.microsoft.com/en-us/azure/storage/blobs/anonymous-read-access-configure?tabs=portal
[46]: https://azure.microsoft.com/fr-fr/updates/choose-to-allow-or-disallow-blob-public-access-on-azure-storage-accounts/

> [47]: https://azure.microsoft.com/fr-fr/updates/choose-to-allow-or-disallow-blob-public-access-on-azure-storage-accounts/

Reviewer: Noting that this link is a duplicate of the one for 46. I think this link is supposed to point elsewhere.
